### DISCUSSION TITLE
Any way to suppress "zsh completion" message?

### DISCUSSION THREAD

**Discussion by johnflavin** [Homebrew/discussions#1539] [Everyday usage]
Some formulae, like `awscli` for example, install shell completions alongside the cli tool. When these formulae are installed or updated they print a message about it:
```
==> Caveats
==> awscli
The "examples" directory has been installed to:
  /usr/local/share/awscli/examples

zsh completions and functions have been installed to:
  /usr/local/share/zsh/site-functions
```

I think that's fine in general and all that, shell completions are super helpful and I suppose having a way to alert the user that they've been installed is nice. But I don't use `zsh`, I use `fish`. So I don't need to know that there have been shell completions installed for a shell I don't use.

Is there any way to suppress the printing of these "zsh completions" messages, or even all the caveats? Any flag / environment variable / config option to set or anything like that?

**Comment by carlocab** [comment]
I think you shouldn't be seeing messages about Zsh completions, actually. On Zsh, I don't see any messages about formulae installed Bash or fish completions.

For example, `deno` installs completions for all three shells but I only see the message about Zsh completion when I do `brew install deno`. Similarly, if I open a bash shell, then `brew install deno` only shows me the message about Bash completions.

It seems to me that either there's a bug in `brew` or it's not detecting your shell correctly. What's the output of `echo $SHELL`? (or whatever the fish equivalent is...)

**Reply by johnflavin** [reply]
Interestingly, it is zsh.
```
> echo $SHELL
/bin/zsh
```

For more context, I haven't set fish as my system default shell using `chsh` because it isn't POSIX compliant. Instead I set it as the default within my terminal application; whenever I open a new terminal it runs `fish`. And I suppose the implication is that fish doesn't set `SHELL`.

I'll try manually setting `SHELL` and installing something...

Yep, if I manually set `SHELL` to the location of fish, then I get this message:
```
==> Caveats
fish completions have been installed to:
  /usr/local/share/fish/vendor_completions.d
```

So it looks like I have identified the source of the issue (fish isn't setting `$SHELL`), a path to fixing it upstream (check with the fish people why this is happening), and a workaround (set `$SHELL` myself). Thanks!

**Reply by carlocab** [reply]
This sounds like expected behaviour.

Chances are is that your terminal first starts Zsh (since that's your default shell) then `exec`s `fish`. By the time this happens, the initial Zsh session will have set `SHELL=/bin/zsh`, and `exec fish` just inherits that env variable.

You can see this in action by setting `SHELL` to whatever value (make sure to `export` it, or whatever the fish equivalent is), then doing `exec bash`. You'll see that the bash session you've started will inherit the previous value of `SHELL` that you've set.

**Reply by tjluoma** [reply]
> For more context, I haven't set fish as my system default shell using chsh because it isn't POSIX compliant. 

I don't understand why that would stop you from setting it as your default shell. (Not saying there isn't a good/valid reason, just saying that I don't know what that would be.)

However, perhaps this suggestion will help: whenever I setup a Mac, I always make a second Administrator account, and leave it with as-default-as possible settings, no apps set to launch at login, etc. The idea is that if I ever do something that majorly messes up my primary account, I'll always have another admin account that I can use to try to fix things. (It's also sometimes helpful to be able to log in to an account and know that nothing is going to autolaunch, especially if I'm trying to do something like run an update, etc. which I know is going to make me reboot again anyway.)

I also make sure that this account uses a different login shell than I use, just in case something happens with my regular shell or its configuration files that prevents me from being able to log in as usual.

So, in your case, you could just create a second admin user and leave it set to `zsh` and then `chsh` your regular account's `$SHELL` to `fish`.

I can't say that I have ever gotten into a situation so dire that I _had to_ use my backup admin account, but I have _chosen_ to use it on some occasions, usually as a diagnostics tool.

FWIW.

**Reply by johnflavin** [reply]
> I don't understand why that would stop you from setting it as your default shell. (Not saying there isn't a good/valid reason, just saying that I don't know what that would be.)

This is a good observation. I do not have a “good” reason, more of a “paranoid” or “anxiety” reason that I don’t know what processes are going to run on the system expecting a posix compliant shell, and I don’t want to break anything out of ignorance. I feel it is achieving a similar goal to the siloed admin account; when I run things the way I always run them, in my preferred apps and so forth, I have things set up in the non-default way I like. But when possible I try not to touch settings outside that context.
